File: auxiliary/scribe/test/live-input.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');

function isAtOrBelow(candidate, root) {
  const relative = path.relative(root, candidate);
  return relative === '' || (!relative.startsWith('..' + path.sep) && relative !== '..' && !path.isAbsolute(relative));
}

function currentProfile() {
  return path.resolve(process.env.USERPROFILE || os.homedir());
}

function assertAccountBoundary(candidate, label) {
  const profile = currentProfile();
  const profilesRoot = path.dirname(profile);
  if (isAtOrBelow(candidate, profilesRoot) && !isAtOrBelow(candidate, profile)) {
    throw new Error(`${label} points into a different user profile.`);
  }
}

function requiredLiveFile(envName) {
  const raw = String(process.env[envName] || '').trim();
  if (!raw) throw new Error(`${envName} is required; no live test input is selected by default.`);
  if (process.platform === 'win32' && /^[\\/]{2}/.test(raw)) {
    throw new Error(`${envName} must not use a UNC or device namespace.`);
  }
  if (!path.isAbsolute(raw)) throw new Error(`${envName} must be an absolute path.`);
  const lexical = path.resolve(raw);
  assertAccountBoundary(lexical, envName);
  const resolved = fs.realpathSync.native(lexical);
  assertAccountBoundary(resolved, envName);
  if (!fs.statSync(resolved).isFile()) throw new Error(`${envName} must name a file.`);
  return resolved;
}

function resolveExecutable(envName, executableNames) {
  if (String(process.env[envName] || '').trim()) return requiredLiveFile(envName);
  for (const directory of String(process.env.PATH || '').split(path.delimiter)) {
    if (!directory) continue;
    for (const name of executableNames) {
      const candidate = path.resolve(directory.replace(/^"|"$/g, ''), name);
      try {
        assertAccountBoundary(candidate, 'PATH executable');
      } catch (_) {
        continue;
      }
      try {
        if (fs.statSync(candidate).isFile()) {
          const resolved = fs.realpathSync.native(candidate);
          assertAccountBoundary(resolved, 'PATH executable');
          return resolved;
        }
      } catch (_) {}
    }
  }
  throw new Error(`${envName} is not set and no matching executable is on PATH.`);
}

module.exports = { requiredLiveFile, resolveExecutable };

File: auxiliary/scribe/test/live_input.py
"""Explicit, account-confined inputs for opt-in live Scribe tests."""

import os
from pathlib import Path


def _at_or_below(candidate, root):
    candidate = os.path.normcase(os.path.abspath(candidate))
    root = os.path.normcase(os.path.abspath(root))
    try:
        return os.path.commonpath((candidate, root)) == root
    except ValueError:
        return False


def _assert_account_boundary(candidate, label):
    profile = os.path.abspath(os.environ.get("USERPROFILE") or str(Path.home()))
    profiles_root = os.path.dirname(profile)
    if _at_or_below(candidate, profiles_root) and not _at_or_below(candidate, profile):
        raise RuntimeError("%s points into a different user profile." % label)


def required_live_file(env_name):
    raw = os.environ.get(env_name, "").strip()
    if not raw:
        raise RuntimeError(
            "%s is required; no live test input is selected by default." % env_name
        )
    if os.name == "nt" and raw.startswith(("\\\\", "//")):
        raise RuntimeError("%s must not use a UNC or device namespace." % env_name)
    if not os.path.isabs(raw):
        raise RuntimeError("%s must be an absolute path." % env_name)
    lexical = os.path.abspath(raw)
    _assert_account_boundary(lexical, env_name)
    resolved = os.path.realpath(lexical)
    _assert_account_boundary(resolved, env_name)
    if not os.path.isfile(resolved):
        raise RuntimeError("%s must name a file." % env_name)
    return Path(resolved)
